refactor!: ADR-038 go-plugin system

egaxhaj opened the pull request:
<!--
The default pull request template is for types feat, fix, or refactor.
For other templates, add one of the following parameters to the url:
- template=docs.md
- template=other.md
-->

## Description

For https://github.com/cosmos/cosmos-sdk/issues/10096

replaces #11691

implements #13473

This is an extension and refactor of the existing ADR-038 streaming service work to introduce a plugin system over gRPC to the SDK and load streaming services using  the [hashicorp/go-plugin](https://github.com/hashicorp/go-plugin) system rather the approach taken in #11691 (which uses Go's built in Plugins API).

The plugin system introduced here is meant to be extensible, so that if other components/features of the SDK wish to be included as plugins in the future they can write plugins with minimal effort by defining an interface and a gRPC message protocol and leveraging the built in plugin system.

Closes: #XXXX

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

julienrbrt commented on streaming/plugins/abci/v1/README.md:
```diff
@@ -0,0 +1,426 @@
+# ABCI and State Streaming Plugin (gRPC)
+
+The `BaseApp` package contains the interface for a [ABCIListener](../../../../baseapp/streaming.go)
```
Let's have direct links from github. Path changes and links get broken.

egaxhaj replied:
added direct links.

julienrbrt commented on streaming/plugins/abci/v1/examples/plugin-python/kafka.py:
```diff
@@ -0,0 +1,97 @@
+from concurrent import futures
```
Do we really need to store the python example in the repo?
We don't have any CI to test that it actually works. I feel like it could be put somewhere else.

egaxhaj replied:
No they don't need to be in there. I'll remove the python code.

egaxhaj commented on baseapp/abci.go:
```diff
@@ -199,10 +201,24 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
+			if err := abciListener.ListenBeginBlock(ctx, req, res); err != nil {
+				app.logger.Error("BeginBlock listening hook failed", "height", blockHeight, "err", err)
+				if app.stopNodeOnABCIListenerErr {
+					plugin.CleanupClients()
+					os.Exit(1)
```
I put this here for feedback because I don't have an idea yet where it should go. I don't like it that the API is aware of the plugin cleanup here. Perhaps fold it into a function in the BaseApp? Thoughts?

=> `app.stopABCIListenersAndExit()`?


@tac0turtle @alexanderbez @peterbourgon

kocubinski replied:
Could the `abci.ListenBeginBlock` implementation not handle both `os.Exit(1)` and cleanup? This would simplify the top level control flow here.

alexanderbez replied:
What are your thoughts @egaxhaj? Seems like pushing this down to `ListenBeginBlock` is cleaner, no?

egaxhaj replied:
Moved down to the plugin.

https://github.com/cosmos/cosmos-sdk/blob/31a9e9c49299bbc65f804d3bf99182da98f43c83/store/streaming/abci/grpc.go#L88

egaxhaj commented on streaming/plugins/proto/abci/v1/grpc.proto:
```diff
@@ -0,0 +1,49 @@
+syntax = "proto3";
+
+package cosmos.sdk.grpc.abci.v1;
```
Having the Proto here feels out of place. It needs to be with the rest of the SDK Proto files. It can easily be pushed up to buf.build and downloaded as well. 

I'm moving it to `proto/sreaming/abci/v1/grpc.proto` and renaming the package to `cosmos.streaming.abci.v1`.

tac0turtle commented on proto/cosmos/streaming/abci/v1/grpc.proto:
```diff
@@ -0,0 +1,60 @@
+syntax = "proto3";
+
+package cosmos.streaming.abci.v1;
```
why is this part of abci? abci should be a minimal layer between the app between modules and tendermint. Im leaning towards not adding more things to it like this

egaxhaj replied:
This is under streaming. I called it abci because its streaming abci events. Where should it go?

tac0turtle replied:
why was this moved out of store? this was moved into store to house all things storage within store. can you please move it back or provide reasoning for this move

egaxhaj replied:
I'm confused. Are you mistaking it  for `proto/comsos/base/store/v1beta1/listening.proto`? This is a brand new grpc service and was never in store. If the above is not correct, please advice on what the package name and location of the proto should be.

egaxhaj replied:
moved package to `store`.

tac0turtle commented on simapp/app.go:
```diff
@@ -263,10 +267,23 @@ func NewSimApp(
-		logger.Error("failed to load state streaming", "err", err)
-		os.Exit(1)
+	// register streaming services
+	streamingCfg := cast.ToStringMap(appOpts.Get(baseapp.StreamingTomlKey))
+	for service := range streamingCfg {
+		pluginKey := fmt.Sprintf("%s.%s.%s", baseapp.StreamingTomlKey, service, baseapp.StreamingABCIPluginTomlKey)
+		pluginName := strings.TrimSpace(cast.ToString(appOpts.Get(pluginKey)))
+		if len(pluginName) > 0 {
+			logLevel := cast.ToString(appOpts.Get(flags.FlagLogLevel))
+			plugin, err := streaming.NewStreamingPlugin(pluginName, logLevel)
+			if err != nil {
+				fmt.Printf("failed to load streaming plugin: %s", err)
+				os.Exit(1)
+			}
+			if err := baseapp.RegisterStreamingPlugin(bApp, appOpts, keys, plugin); err != nil {
+				fmt.Printf("failed to register streaming plugin: %s", err)
+				os.Exit(1)
+			}
+		}
```
this should be cleaned up. there can be a helper function that does this

egaxhaj replied:
moved it into function in `test_helpers.go`

tac0turtle commented on streaming/README.md:
```diff
@@ -0,0 +1,27 @@
+# Cosmos-SDK Plugins
```
please don't put this in the root, we moved it to store because we are trying to house all storage related items in one place

egaxhaj replied:
it now resides in `store/streaming`

tac0turtle reviewed (commented):
left some initial comments. there are some changes that should be undone

tac0turtle commented on baseapp/abci.go:
```diff
@@ -10,17 +10,19 @@ import (
-	abci "github.com/tendermint/tendermint/abci/types"
-	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	"github.com/hashicorp/go-plugin"
 	"google.golang.org/grpc/codes"
 	grpcstatus "google.golang.org/grpc/status"
 
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+
```
please group external imports

egaxhaj replied:
fixed.

kocubinski commented on baseapp/streaming.go:
```diff
@@ -0,0 +1,105 @@
+	exposedKeys := exposeStoreKeysSorted(exposeKeysStr, keys)
+	bApp.cms.AddListeners(exposedKeys)
+	bApp.SetStreamingService(abciListener)
+	bApp.stopNodeOnABCIListenerErr = stopNodeOnErr
```
I am skeptical that BaseApp is the best place for `stopNodeOnABCIListenerErr`, since it leads to the kind of [question you raised](https://github.com/cosmos/cosmos-sdk/pull/14207/files#r1043726514).  

I propose passing this configuration state into the listener instead of BaseApp.  What do you think?

egaxhaj replied:
I'm resolving this conversation. It continues in https://github.com/cosmos/cosmos-sdk/pull/14207#discussion_r1067564955

alexanderbez commented on baseapp/baseapp.go:
```diff
@@ -142,7 +142,14 @@ type BaseApp struct { //nolint: maligned
 	// abciListeners for hooking into the ABCI message processing of the BaseApp
 	// and exposing the requests and responses to external consumers
-	abciListeners []storetypes.ABCIListener
+	abciListeners []ABCIListener
+
+	// abciListenersAsync for determining if abciListeners will run asynchronously.
+	abciListenersAsync bool
+
+	// stopNodeOnABCIListenerErr halts the node when ABCI streaming service listening results in an error.
+	// stopNodeOnABCIListenerErr=true MUST be paired with abciListenersAsync=false, otherwise it will be ignored.
+	stopNodeOnABCIListenerErr bool
```
Why not have this all be part of a single `StreamingManager` type? This manager could be responsible for all writes and cleanup.

egaxhaj replied:
I agree, I'm looking at the best way to push it down to the `ListenBeginBlock`. If you take a look the `NewStreamingPlugin ` function, it serves plugins by doing a lookup through the `PluginMap`, which is a struct, and the function itself is `NewStreamingPlugin`. One way I'm looking at is to pass it down through the `ctx`. This would require updating `setState` (see below) and perhaps adding a `StreamingManager` function to the `sdk.Context{}`?


https://github.com/cosmos/cosmos-sdk/blob/3ed0fedb509749c14f96f71cf4ba3ce031539314/streaming/streaming.go#L23-L31

What I'm currently testing with the below. It needs to be set for every new context.
```go
func (app *BaseApp) setState(mode runTxMode, header tmproto.Header) {
	...
	baseState.ctx = baseState.ctx.WithValue("stopNodeOnABCIListenerErr", app.stopNodeOnABCIListenerErr)
        ...
```

perhaps the ^^^ could become?

```go
func (app *BaseApp) setState(mode runTxMode, header tmproto.Header) {
	...
        baseState.ctx = baseState.ctx.WithStreamingManager(app.streamingManager)
        ...
```

egaxhaj replied:
this relates to the question from @kocubinski as well https://github.com/cosmos/cosmos-sdk/pull/14207#discussion_r1066025287.

egaxhaj replied:
@alexanderbez @kocubinski take a look at it now and let me know what you think. I've grouped them into a `StreamingManager` and moved the cleanup and exit to the plugin layer.

kocubinski replied:
Personally I find this approach has a better separation of concerns :clap:

tac0turtle commented on store/streaming/plugins/abci/v1/grpc.go:
```diff
@@ -0,0 +1,148 @@
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
```
store cant depend on sdk could we remove this import

egaxhaj replied:
The plugin needs to `UnwrapSDKContext` to get access to the `BlockHeight` and the `StreamingManager`. We need to make a compromise here if we are to house streaming under store, otherwise, we move streaming back to root. Let me know which takes precedence and I'll make the necessary changes.

tac0turtle replied:
neither is really a good option imo. we have purposefully removed sdk from store and moved streaming to store. There is a way around this. We just have to figure it out

egaxhaj replied:
To remove the sdk dependency I introduced a `Context` interface. This interface is implemented by `types.Context` in the SDK. With store being a standalone module, I found introducing an interface to pass context information to be a good approach here. Would like some feedback on this.

https://github.com/cosmos/cosmos-sdk/blob/31a9e9c49299bbc65f804d3bf99182da98f43c83/store/types/context.go#L8-L15

implementation in sdk:
https://github.com/cosmos/cosmos-sdk/blob/31a9e9c49299bbc65f804d3bf99182da98f43c83/types/context.go#L313-L314

use in plugin:
https://github.com/cosmos/cosmos-sdk/blob/31a9e9c49299bbc65f804d3bf99182da98f43c83/store/streaming/abci/grpc.go#L20-L26

tac0turtle commented on store/streaming/plugins/abci/v1/README.md:
```diff
@@ -0,0 +1,396 @@
+# ABCI and State Streaming Plugin (gRPC)
```
what is the reasoning to have the import path be plugins/abci/v1? can we do only streaming since we expect the version in the sdk to only be plugin based

egaxhaj replied:
The plugin system uses a handshake to make sure the correct plugin version is loaded. So it's important we version the ABCI listener service for future updates. I am also keeping in line with SDKs proto versioning scheme (although v1 may not be exactly correct).

Versioning may pan out not to be necessary in this case and we can remove it down the road but I don't see the harm in at least starting down this path. Ultimately, this is an SDK decision. I'll change it to what you decide but at a minimum we need to keep `streaming/abci` to distinguish between different plugins. 

plugin handshakes.
 https://github.com/cosmos/cosmos-sdk/blob/a36f16565381a2cf7deb7237e48ddaef8d96b7e6/store/streaming/streaming.go#L18-L20

tac0turtle replied:
what other sorts plugins do you envision? semantic import could cause confusion with golang semantic confusion.

egaxhaj replied:
Say that in the future you wanted to track some information regarding smart contracts. That could be a plugin `streaming/wasmd` etc. As to the semantic versioning I have no problem removing it and reducing to `streaming/abci`. Reducing it to `streaming/` would cause confusion in its own because the plugin system is built to support different plugins (protocols).

egaxhaj replied:
Ready to fold this to `streaming/abci`?

tac0turtle replied:
yea that works, thanks

egaxhaj replied:
completed.

peterbourgon commented on store/streaming/plugins/abci/v1/grpc.go:
```diff
@@ -0,0 +1,148 @@
+	if sm.Async {
+		go func(ctx sdk.Context) {
+			_, err := m.client.ListenEndBlock(ctx, request)
+			ctx.Logger().Error("EndBlock listening hook failed", "height", ctx.BlockHeight(), "err", err)
```
This should not execute if err == nil, but at the moment it does.

peterbourgon commented on store/streaming/plugins/abci/v1/grpc.go:
```diff
@@ -0,0 +1,148 @@
+	if sm.Async {
+		go func(ctx sdk.Context) {
+			_, err := m.client.ListenBeginBlock(ctx, request)
+			ctx.Logger().Error("BeginBlock listening hook failed", "height", ctx.BlockHeight(), "err", err)
```
This should not execute if err == nil, but at the moment it does.

egaxhaj replied:
Ah. I missed the check. 

I know we decided to add goroutines here for fire-and-forget but would it not be better to leave this up to the plugin implementation? This would also satisfy CI non-determinism checks (which it flagged). We can add a ctx `timeout` to mitigate any blocking concerns?

peterbourgon replied:
If you want to leave this up to the implementation then you would need to remove the Async flag from the sm struct and always call ListenBeginBlock directly. In either case you should only call logger.Error if the err returned by the client call is non-nil.

egaxhaj replied:
`async` functionality has been removed.

peterbourgon commented on store/streaming/plugins/abci/v1/grpc.go:
```diff
@@ -0,0 +1,148 @@
+	Impl storetypes.ABCIListener
+}
+
+func (m GRPCServer) ListenBeginBlock(ctx context.Context, request *ListenBeginBlockRequest) (*ListenBeginBlockResponse, error) {
+	if err := m.Impl.ListenBeginBlock(ctx, *request.Req, *request.Res); err != nil {
+		return nil, err
+	}
+	return &ListenBeginBlockResponse{}, nil
+}
+
+func (m GRPCServer) ListenEndBlock(ctx context.Context, request *ListenEndBlockRequest) (*ListenEndBlockResponse, error) {
+	if err := m.Impl.ListenEndBlock(ctx, *request.Req, *request.Res); err != nil {
+		return nil, err
+	}
+	return &ListenEndBlockResponse{}, nil
+}
+
+func (m GRPCServer) ListenDeliverTx(ctx context.Context, request *ListenDeliverTxRequest) (*ListenDeliverTxResponse, error) {
+	if err := m.Impl.ListenDeliverTx(ctx, *request.Req, *request.Res); err != nil {
+		return nil, err
+	}
+	return &ListenDeliverTxResponse{}, nil
+}
+
+func (m GRPCServer) ListenCommit(ctx context.Context, request *ListenCommitRequest) (*ListenCommitResponse, error) {
+	if err := m.Impl.ListenCommit(ctx, *request.Res, request.ChangeSet); err != nil {
+		return nil, err
+	}
+	return &ListenCommitResponse{}, nil
+}
```
All of these methods return empty responses on success, is that intentional?

egaxhaj replied:
Yes, I don't see a reason to report additional information upstream.

egaxhaj replied:
Is this acceptable? Can I resolve this conversation?

egaxhaj commented:
The Dependency Review `required` job is failing. I looked to see if `govulnchbeck` allows for skipping certain checks but it doesn't. Any recommendations on what we should do with the examples?

```shell
govulncheck: Packages contain errors:
Error: /home/runner/work/cosmos-sdk/cosmos-sdk/store/streaming/plugins/abci/v1/examples/stdout.go:45:6: main redeclared in this block
Error: /home/runner/work/cosmos-sdk/cosmos-sdk/store/streaming/plugins/abci/v1/examples/file.go:99:6: 	other declaration of main
```

@tac0turtle @peterbourgon @kocubinski @alexanderbez

alexanderbez commented:
I don't follow? Can't those be resolved?

peterbourgon commented on store/rootmulti/store.go:
```diff
@@ -401,23 +401,36 @@ func (rs *Store) TracingEnabled() bool {
-		rs.listeners[key] = append(ls, listeners...)
-	} else {
-		rs.listeners[key] = listeners
+// AddListeners adds state change listener for a specific KVStore
+func (rs *Store) AddListeners(keys []types.StoreKey) {
+	listener := types.NewMemoryListener()
+	for i := range keys {
+		rs.listeners[keys[i]] = listener
```
This doesn't add a listener for provided keys, it overwrites whatever listener exists for those keys with a brand new and empty listener, which is a destructive (and breaking) change.

egaxhaj replied:
fixed to add listeners only if one doesn't exist for that key.

peterbourgon commented on store/types/store.go:
```diff
@@ -194,9 +194,11 @@ type CommitMultiStore interface {
+	// AddListeners adds q listener for the KVStore belonging to the provided StoreKey
+	AddListeners(keys []StoreKey)
+
+	// PopStateCache returns the accumulated state change messages from MemoryListener
+	PopStateCache() []*StoreKVPair
```
1. It returns the state changes messages accumulated from the _CommitMultiStore_ which the method is implemented by
2. It _destroys_ those state change messages in its implementation, which is a mutating operation, that isn't described in documentation or synchronized in any way

egaxhaj replied:
added comments to implementation and synchronized method.

peterbourgon commented on baseapp/abci.go:
```diff
@@ -426,10 +437,13 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
-		if err := streamingListener.ListenCommit(app.deliverState.ctx, res); err != nil {
-			panic(fmt.Errorf("Commit listening hook failed, height: %d, err: %w", header.Height, err))
+	// call the streaming service hook with the EndBlock messages
+	for _, abciListener := range app.streamingManager.AbciListeners {
+		ctx := app.deliverState.ctx
+		blockHeight := ctx.BlockHeight()
+		changeSet := app.cms.PopStateCache()
+		if err := abciListener.ListenCommit(ctx, res, changeSet); err != nil {
+			app.logger.Error("Commit listening hook failed", "height", blockHeight, "err", err)
 		}
```
You're calling app.cms.PopStateCache for every AbciListener, but PopStateCache is a destructive operation, which invalidates the returned value of the app.cms. This means that only the first streamingManager.AbciListener will get a valid changeSet.

Where is this method tested? How did this bug survive to this PR?

egaxhaj replied:
Looks like we all missed it during ADR review. I added unit testing and fixed the call so all listeners receive a valid changeSet.

egaxhaj replied:
Going to take a second look and improve it a bit

egaxhaj replied:
done.

peterbourgon commented on store/listenkv/store_test.go:
```diff
@@ -149,13 +140,12 @@ func TestListenKVStoreDelete(t *testing.T) {
 		store.Delete(tc.key)
-		storeKVPair := new(types.StoreKVPair)
-		testMarshaller.UnmarshalLengthPrefixed(buf.Bytes(), storeKVPair)
+		cache := listener.PopStateCache()
+		storeKVPair := cache[len(cache)-1]
```
How do you know that PopStateCache returns a slice of length > 0?

egaxhaj replied:
fixed. added `require.NotEmpty(t, cache)` one line above.

peterbourgon commented on store/streaming/plugins/abci/v1/README.md:
```diff
@@ -0,0 +1,391 @@
+	client ABCIListenerServiceClient
+}
+
+func (m GRPCClient) ListenBeginBlock(ctx context.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error {
+	_, err := m.client.ListenBeginBlock(ctx, &ListenBeginBlockRequest{
+		Req: &req,
+		Res: &res,
+	})
+	return err
+}
+
+func (m GRPCClient) ListenEndBlock(ctx context.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) error {
+	_, err := m.client.ListenEndBlock(ctx, &ListenEndBlockRequest{
+		Req: &req,
+		Res: &res,
+	})
+	return err
+}
```
But wait, these Listen methods mutate stuff in the Res field they're given, right? But the res provided to this client's ListenWhateverBlock is a value type, which means even if you call client.ListenWhatever with a `&res` address of that input param, whatever changes the ListenWhateverBlock method(s) make to that parameter will not be visible to the caller of this GRPCClient method. Those changes are applied to the _copy of_ the `res` value passed to the method only.

This is a pretty straightforward bug that should have been caught by unit tests -- where is this method tested?

egaxhaj replied:
Why would we want to make changes to the req, res?

SpicyLemon replied:
> But wait, these Listen methods mutate stuff in the Res field they're given, right? But the res provided to this client's ListenWhateverBlock is a value type, which means even if you call client.ListenWhatever with a `&res` address of that input param, whatever changes the ListenWhateverBlock method(s) make to that parameter will not be visible to the caller of this GRPCClient method. Those changes are applied to the _copy of_ the `res` value passed to the method only.
> 
> This is a pretty straightforward bug that should have been caught by unit tests -- where is this method tested?

These are listeners; they shouldn't be changing anything. And as you pointed out, they can't.

peterbourgon replied:
If they're not changing anything, then it's confusing that they take req and res by reference, given the outer method here receives them by value. But there's probably an obscure reason for it, I guess.

egaxhaj replied:
All Go generated gRPC code uses references for the client/server methods. I'm guessing the reason has to do with protoreflection. [#3 on this blog post](https://go.dev/blog/laws-of-reflection) may shed some light to the reason.

peterbourgon commented:
As it stands this PR is definitely broken and cannot be merged.

peterbourgon commented on store/rootmulti/store.go:
```diff
@@ -77,6 +77,8 @@ type Store struct {
 
 	metrics metrics.StoreMetrics
+
+	mtx sync.Mutex
```
What is the purpose of this mutex? What is is meant to guard? At the moment you lock it in PopStateCache, to read rs.listeners, but in that case you also need to lock it in any other method that accesses rs.listeners, for example AddListeners.

egaxhaj replied:
updated to lock calls to rs.listeners.

egaxhaj commented:
ref: https://github.com/cosmos/cosmos-sdk/pull/14207#issuecomment-1399125285

> I don't follow? Can't those be resolved?

@alexanderbez My question exactly. How would we resolve it? CI is picking up the `main` method in the Plugin.

I don't want to remove them. It's good to have examples as a reference.

peterbourgon commented on store/rootmulti/store.go:
```diff
@@ -401,23 +403,43 @@ func (rs *Store) TracingEnabled() bool {
 	return false
 }
 
+// PopStateCache returns the accumulated state change messages from the CommitMultiStore
+// Calling PopStateCache destroys the currently accumulated state. This is a mutating
+// and destructive operation. This method has been synchronized.
```
It destroys the accumulated state in each listener, but not in the store itself?

egaxhaj replied:
Ah. Yes. Updated to the following.

>// PopStateCache returns the accumulated state change messages from the CommitMultiStore
// Calling PopStateCache destroys only the currently accumulated state in each listener
// not the state in the store itself. This is a mutating and destructive operation.
// This method has been synchronized.

peterbourgon commented on store/rootmulti/store.go:
```diff
@@ -401,23 +403,43 @@ func (rs *Store) TracingEnabled() bool {
+	defer rs.mtx.Unlock()
+	var cache []*types.StoreKVPair
+	for key := range rs.listeners {
+		ls := rs.listeners[key]
+		if ls != nil {
+			cache = append(cache, ls.PopStateCache()...)
+		}
```
Does each listener not receive the same state?

egaxhaj replied:
Each listener is assigned it's own `MemoryListener` in `AddListeners` so we need to collect state from all of them.

peterbourgon commented on store/streaming/abci/examples/file/file.go:
```diff
@@ -0,0 +1,109 @@
+		return err
+	}
+
+	filename := fmt.Sprintf("%s/%s.txt", home, file)
+	f, err := os.OpenFile(filepath.Clean(filename), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		return err
+	}
+	a.fileDict[file] = f
+
+	if _, err := f.Write(data); err != nil {
+		return err
+	}
+
+	return nil
```
Where are these open file descriptors closed?

egaxhaj replied:
fixed.

peterbourgon commented on store/streaming/abci/examples/stdout/stdout.go:
```diff
@@ -0,0 +1,55 @@
+	BlockHeight int64
+}
+
+func (a StdoutPlugin) ListenBeginBlock(ctx context.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error {
+	a.BlockHeight = req.Header.Height
+	// process tx messages (i.e: sent to external system)
+	fmt.Printf("listen-begin-block: block-height=%d req=%v res=%v", a.BlockHeight, req, res)
+	return nil
+}
```
This method is defined on a value receiver, so setting a.BlockHeight will only apply to this function call, it won't be visible to other consumers.

egaxhaj replied:
fixed.

peterbourgon commented on store/rootmulti/store.go:
```diff
@@ -398,23 +400,46 @@ func (rs *Store) TracingEnabled() bool {
 
 // ListeningEnabled returns if listening is enabled for a specific KVStore
 func (rs *Store) ListeningEnabled(key types.StoreKey) bool {
 	if ls, ok := rs.listeners[key]; ok {
-		return len(ls) != 0
+		return ls != nil
 	}
 	return false
 }
```
This reads rs.listeners without taking the listenersMx.

egaxhaj replied:
The listenersMx is being applied on the callers. They call `ListeningEnabled` inside a loop. That's why I left it out.

https://github.com/cosmos/cosmos-sdk/blob/be35b4fdceeaf889471d89f5a4667482ef1aedf1/store/rootmulti/store.go#L512

https://github.com/cosmos/cosmos-sdk/blob/be35b4fdceeaf889471d89f5a4667482ef1aedf1/store/rootmulti/store.go#L532

https://github.com/cosmos/cosmos-sdk/blob/be35b4fdceeaf889471d89f5a4667482ef1aedf1/store/rootmulti/store.go#L588

peterbourgon replied:
Then to make sure nothing else can call ListeningEnabled without first taking that mutex, it should probably be un-exported e.g. `listeningEnabled`.

egaxhaj replied:
changed to un-exported.

tac0turtle commented on simapp/app_v2.go:
```diff
@@ -245,9 +249,23 @@ func NewSimApp(
-		logger.Error("failed to load state streaming", "err", err)
-		os.Exit(1)
+	// register streaming services
+	streamingCfg := cast.ToStringMap(appOpts.Get(baseapp.StreamingTomlKey))
+	for service := range streamingCfg {
+		pluginKey := fmt.Sprintf("%s.%s.%s", baseapp.StreamingTomlKey, service, baseapp.StreamingABCIPluginTomlKey)
+		pluginName := strings.TrimSpace(cast.ToString(appOpts.Get(pluginKey)))
+		if len(pluginName) > 0 {
+			logLevel := cast.ToString(appOpts.Get(flags.FlagLogLevel))
+			plugin, err := streaming.NewStreamingPlugin(pluginName, logLevel)
+			if err != nil {
+				fmt.Printf("failed to load streaming plugin: %s", err)
+				os.Exit(1)
+			}
+			if err := baseapp.RegisterStreamingPlugin(app.BaseApp, appOpts, app.kvStoreKeys(), plugin); err != nil {
+				fmt.Printf("failed to register streaming plugin: %s", err)
+				os.Exit(1)
+			}
+		}
```
can we pull this out into a util function in types or a different location. We should avoid adding more boiler plate to the app.go's

egaxhaj replied:
moved this in `baseapp/streaming.go`

tac0turtle commented on store/rootmulti/store.go:
```diff
@@ -74,9 +74,11 @@ type Store struct {
 
 	metrics metrics.StoreMetrics
+
+	listenersMx sync.Mutex
```
can we add a godoc what this mutex does? Slightly lost why its on existing methods?

egaxhaj replied:
added comments.

tac0turtle commented on store/types/context.go:
```diff
@@ -0,0 +1,15 @@
+
+// Context is an interface used by an App to pass context information
+// needed to process store streaming requests.
+type Context interface {
```
❤️

tac0turtle commented on types/context.go:
```diff
@@ -41,29 +41,31 @@ type Context struct {
+func (c Context) Priority() int64                               { return c.priority }
+func (c Context) KVGasConfig() storetypes.GasConfig             { return c.kvGasConfig }
+func (c Context) TransientKVGasConfig() storetypes.GasConfig    { return c.transientKVGasConfig }
+func (c Context) StreamingManager() storetypes.StreamingManager { return c.streamingManager }
```
if this is only being used by store should we pass it directly to the store instead of adding it to the context?

egaxhaj replied:
Do you have an idea in mind? By having it here I was able to introduce the Context interface in the store (removing the sdk dependency) and having types/context implement the store Context interface to do a type conversion in the ABCIListener methods and retrieve the StreamingManager.

tac0turtle replied:
lets leave it for now. In the future this will change, but no need to block this pr on that. Thank you

tac0turtle reviewed (commented):
did a first pass, sorry for being slow on review. Left one question on the mutex that was added to store. Would be good to get some benchmarks to see if it affects anything. 

Thank you for the well written documentation!!

tac0turtle commented:
changes look good, it would be good to understand how big of a performance hit there is so we can communicate that to users. There might be a few benchmarks in store that we can use

julienrbrt commented on go.mod:
```diff
@@ -164,6 +170,9 @@ replace (
 
 // Below are the long-lived replace of the Cosmos SDK
 replace (
+	// TODO tag after https://github.com/cosmos/cosmos-sdk/pull/14207 merge
```
Small nit, these replace are in the wrong group.

egaxhaj replied:
fixed

julienrbrt commented on tests/go.mod:
```diff
@@ -200,6 +205,8 @@ replace (
 replace (
 	// We always want to test against the latest version of the simapp.
 	cosmossdk.io/simapp => ../simapp
+	// TODO tag after https://github.com/cosmos/cosmos-sdk/pull/14207 merge
```
ditto

egaxhaj replied:
fixed

julienrbrt commented on simapp/go.mod:
```diff
@@ -203,6 +208,8 @@ replace (
 
 // Below are the long-lived replace of the SimApp
 replace (
+	// TODO tag after https://github.com/cosmos/cosmos-sdk/pull/14207 merge
```
ditto

egaxhaj replied:
fixed.

julienrbrt commented on server/config/config.go:
```diff
@@ -212,8 +192,7 @@ type Config struct {
 	StateSync StateSyncConfig  `mapstructure:"state-sync"`
-	Store     StoreConfig      `mapstructure:"store"`
-	Streamers StreamersConfig  `mapstructure:"streamers"`
+	Streaming StreamingConfig  `mapstructure:"streaming"`
```
A follow-up, or this PR needs to update with how look like the new configuration: https://github.com/cosmos/cosmos-sdk/blob/main/tools/confix/data/v0.48-app.toml
Possibly, we need to add something in the UPGRADING.md as well.

egaxhaj replied:
I updated v048-app.tom. 

I'm certain on the process of UPGRADING.md but if you make a suggestion I can commit it.

julienrbrt commented on simapp/sim_test.go:
```diff
@@ -41,9 +43,14 @@ import (
 // SimAppChainID hardcoded chainID for simulation
 const SimAppChainID = "simulation-app"
 
+var (
```
nit: this could be one line

egaxhaj replied:
fixed.

egaxhaj commented:
> changes look good, it would be good to understand how big of a performance hit there is so we can communicate that to users. There might be a few benchmarks in store that we can use

I'll look at running the `test-sim-benchmark` and post my results.

ryanchristo commented on store/streaming/README.md:
```diff
@@ -1,88 +1,27 @@
-[ADR-038](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-038-state-listening.md)
-and defined in [types/streaming.go](https://github.com/cosmos/cosmos-sdk/blob/main/store/types/streaming.go).
-The child directories contain the implementations for specific output destinations.
+This package contains an extensible plugin system for the Cosmos-SDK. The plugin system leverages the [hashicorp/go-plugin](https://github.com/hashicorp/go-plugin) system. This system is designed to work over RCP. 
```
```suggestion
This package contains an extensible plugin system for the Cosmos-SDK. The plugin system leverages the [hashicorp/go-plugin](https://github.com/hashicorp/go-plugin) system. This system is designed to work over RPC. 
```

ryanchristo replied:
```suggestion
This package contains an extensible plugin system for the Cosmos-SDK. The plugin system leverages the [hashicorp/go-plugin](https://github.com/hashicorp/go-plugin) system. This system is designed to work over RPC. 
```

ryanchristo commented on store/streaming/README.md:
```diff
@@ -1,88 +1,27 @@
-Currently, a `StreamingService` implementation that writes state changes out to
-files is supported, in the future support for additional output destinations can
-be added.
+Although the `go-plugin` is built to work over RCP, it is currently only designed to work over a local network.
```
```suggestion
Although the `go-plugin` is built to work over RPC, it is currently only designed to work over a local network.
```

egaxhaj commented:
> > changes look good, it would be good to understand how big of a performance hit there is so we can communicate that to users. There might be a few benchmarks in store that we can use
> 
> I'll look at running the `test-sim-benchmark` and post my results.


Local benchmark testing with and without streaming enabled. I used the file plugin from the examples: `store/streaming/abci/examples/file/file`.

- `make test-sim-benchmark`
- `make test-sim-profile`
- `make test-sim-benchmark-streaming`
- `make test-sim-profile-streaming`

benchmark (no streaming):

```
Running application benchmark for numBlocks=500, blockSize=200. This may take awhile!

LevelDB Stats
Compactions
 Level |   Tables   |    Size(MB)   |    Time(sec)  |    Read(MB)   |   Write(MB)
-------+------------+---------------+---------------+---------------+---------------
   0   |          2 |       5.29184 |      11.73880 |       0.00000 |     347.31901
   1   |         53 |      98.29939 |     229.59947 |    8560.64045 |    8545.90093
   2   |        116 |     227.99740 |      17.86907 |     656.58859 |     655.59772
-------+------------+---------------+---------------+---------------+---------------
 Total |        171 |     331.58863 |     259.20734 |    9217.22904 |    9548.81767

LevelDB cached block size 6537364
goos: darwin
goarch: amd64
pkg: cosmossdk.io/simapp
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkFullAppSimulation-12    	       1	295792074002 ns/op	87070234456 B/op	1112278620 allocs/op
PASS
ok  	cosmossdk.io/simapp	300.625s
```


profile (no streaming):

```
Running application benchmark for numBlocks=500, blockSize=200. This may take awhile!

LevelDB Stats
Compactions
 Level |   Tables   |    Size(MB)   |    Time(sec)  |    Read(MB)   |   Write(MB)
-------+------------+---------------+---------------+---------------+---------------
   0   |          2 |       5.29183 |      11.33540 |       0.00000 |     347.31867
   1   |         52 |      98.21147 |     220.36103 |    8478.65022 |    8463.93661
   2   |        117 |     228.14920 |      16.78900 |     646.92550 |     645.97293
-------+------------+---------------+---------------+---------------+---------------
 Total |        171 |     331.65249 |     248.48543 |    9125.57571 |    9457.22820

LevelDB cached block size 6388151
goos: darwin
goarch: amd64
pkg: cosmossdk.io/simapp
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkFullAppSimulation-12    	       1	283546864750 ns/op	87250545856 B/op	1112857747 allocs/op
PASS
ok  	cosmossdk.io/simapp	286.361s
```


benchmark (streaming to file - uses the file plugin in the `store/streaming/abci/examples/file/file.go`):

```
Running application benchmark for numBlocks=500, blockSize=200. This may take awhile!

LevelDB Stats
Compactions
 Level |   Tables   |    Size(MB)   |    Time(sec)  |    Read(MB)   |   Write(MB)
-------+------------+---------------+---------------+---------------+---------------
   0   |          2 |       5.29182 |      12.08377 |       0.00000 |     347.31930
   1   |         53 |      98.12084 |     225.45326 |    8676.10402 |    8661.31832
   2   |        116 |     228.17313 |      18.73294 |     705.99500 |     705.04717
-------+------------+---------------+---------------+---------------+---------------
 Total |        171 |     331.58578 |     256.26997 |    9382.09902 |    9713.68480

LevelDB cached block size 6324264
goos: darwin
goarch: amd64
pkg: cosmossdk.io/simapp
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkFullAppSimulation-12    	       1	293033899073 ns/op	87568069136 B/op	1113662536 allocs/op
PASS
ok  	cosmossdk.io/simapp	295.118s
```


profile (streaming to file - uses the file plugin in the `store/streaming/abci/examples/file/file.go`):

```
Running application benchmark for numBlocks=500, blockSize=200. This may take awhile!


LevelDB Stats
Compactions
 Level |   Tables   |    Size(MB)   |    Time(sec)  |    Read(MB)   |   Write(MB)
-------+------------+---------------+---------------+---------------+---------------
   0   |          2 |       5.29181 |      12.50685 |       0.00000 |     347.31966
   1   |         56 |      99.38999 |     243.91219 |    8770.55951 |    8755.79831
   2   |        114 |     226.96492 |      20.49460 |     721.00425 |     720.09249
-------+------------+---------------+---------------+---------------+---------------
 Total |        172 |     331.64672 |     276.91364 |    9491.56375 |    9823.21047

LevelDB cached block size 6372156
goos: darwin
goarch: amd64
pkg: cosmossdk.io/simapp
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkFullAppSimulation-12              1 333175967646 ns/op   87643613136 B/op  1113350970 allocs/op
PASS
ok    cosmossdk.io/simapp  336.236s

tac0turtle reviewed (approved):
utACK. id like to get @alexanderbez and @kocubinski approvals before we merge

tac0turtle commented:
> Local benchmark testing with and without streaming enabled. I used the file plugin from the examples: `store/streaming/abci/examples/file/file`.
> 
> * `make test-sim-benchmark`
> * `make test-sim-profile`
> * `make test-sim-benchmark-streaming`
> * `make test-sim-profile-streaming`
> 
> benchmark (no streaming):

is there a difference to main vs this branch? wondering if the added mutexs would change anything

egaxhaj commented:
@tac0turtle It would be good to communicate pruning configuration options as well when streaming is enabled for non-archival nodes.

egaxhaj commented:
@alexanderbez and @kocubinski can you take another look and merge if all changes look good?

alexanderbez commented on baseapp/abci.go:
```diff
@@ -405,13 +413,15 @@ func (app *BaseApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDeliv
 		return sdkerrors.ResponseDeliverTxWithEvents(err, gInfo.GasWanted, gInfo.GasUsed, sdk.MarkEventsToIndex(anteEvents, app.indexEvents), app.trace)
 	}
 
-	return abci.ResponseDeliverTx{
+	res = abci.ResponseDeliverTx{
```
nit: undo

egaxhaj replied:
not clear here. this is currently `res = abci.ResponseDeliverTx`

alexanderbez commented on baseapp/baseapp.go:
```diff
@@ -414,6 +413,7 @@ func (app *BaseApp) setState(mode runTxMode, header cmtproto.Header) {
 		ms:  ms,
 		ctx: sdk.NewContext(ms, header, false, app.logger),
 	}
+	baseState.ctx = baseState.ctx.WithStreamingManager(app.streamingManager)
```
Why not just move the `WithStreamingManager(app.streamingManager)` to line 414

egaxhaj replied:
moved to line 414.

alexanderbez commented on baseapp/streaming.go:
```diff
@@ -0,0 +1,111 @@
+			plugin, err := streaming.NewStreamingPlugin(pluginName, logLevel)
+			if err != nil {
+				app.logger.Error("failed to load streaming plugin", "error", err)
+				os.Exit(1)
```
Any reason for `os.Exit` over a panic with the error?

egaxhaj replied:
Yes. We've had a conversation about this (ref: https://github.com/cosmos/cosmos-sdk/pull/13473#discussion_r1021917395). Note that If `panic` will ensure that the node will stop,  I would much rather send the error in a panic and clean this up.

kocubinski replied:
Either panic or os.exit seem misplaced.  Why not return the error and let the calling decide what to do?

alexanderbez commented on baseapp/streaming.go:
```diff
@@ -0,0 +1,111 @@
+	keys map[string]*storetypes.KVStoreKey,
+	streamingPlugin interface{},
+) error {
+	switch t := streamingPlugin.(type) {
+	case storetypes.ABCIListener:
+		app.registerABCIListenerPlugin(appOpts, keys, t)
+	default:
+		return fmt.Errorf("unexpected plugin type %T", t)
+	}
+	return nil
```
This would read easier w/o the switch, i.e.

```go
v, ok := streamingPlugin.(storetypes.ABCIListener)
if !ok {
  // ...
}

app.registerABCIListenerPlugin(appOpts, keys, v)
return nil
```

egaxhaj replied:
changed to above.

alexanderbez commented on baseapp/streaming.go:
```diff
@@ -0,0 +1,111 @@
+	return nil
+}
+
+func (app *BaseApp) registerABCIListenerPlugin(
```
This is pretty gnarly code. I would definitely add a godoc documenting what this does. There seems to be some overlap with the logic in `RegisterStreamingServices` too.

egaxhaj replied:
added comment that now reads 
`// registerABCIListenerPlugin registers plugins that implement the ABCIListener interface.`

alexanderbez commented on server/config/toml.go:
```diff
@@ -199,27 +199,30 @@ snapshot-interval = {{ .StateSync.SnapshotInterval }}
-stop-node-on-error = "{{ .Streamers.File.StopNodeOnError }}"
+# The plugin name used for streaming via gRPC.
+# Streaming is only enabled if this is set.
+# Supported plugins: abci_v1
```
What is `abci_v1` plugin?

egaxhaj replied:
set this to be `abci`.

alexanderbez commented on simapp/sim_bench_test.go:
```diff
@@ -38,9 +47,23 @@ func BenchmarkFullAppSimulation(b *testing.B) {
-	appOptions := make(simtestutil.AppOptionsMap, 0)
-	appOptions[flags.FlagHome] = DefaultNodeHome
-	appOptions[server.FlagInvCheckPeriod] = simcli.FlagPeriodValue
+	//
+	//appOptions := make(simtestutil.AppOptionsMap, 0)
+	//appOptions[flags.FlagHome] = DefaultNodeHome
+	//appOptions[server.FlagInvCheckPeriod] = simcli.FlagPeriodValue
```
```suggestion
```

egaxhaj replied:
fixed

alexanderbez commented on store/rootmulti/store.go:
```diff
@@ -485,12 +515,14 @@ func (rs *Store) CacheWrapWithTrace(_ io.Writer, _ types.TraceContext) types.Cac
 // CacheMultiStore creates ephemeral branch of the multi-store and returns a CacheMultiStore.
 // It implements the MultiStore interface.
 func (rs *Store) CacheMultiStore() types.CacheMultiStore {
+	rs.listenersMx.Lock()
+	defer rs.listenersMx.Unlock()
 	stores := make(map[types.StoreKey]types.CacheWrapper)
 	for k, v := range rs.stores {
 		store := types.KVStore(v)
 		// Wire the listenkv.Store to allow listeners to observe the writes from the cache store,
 		// set same listeners on cache store will observe duplicated writes.
-		if rs.ListeningEnabled(k) {
+		if rs.listeningEnabled(k) {
```
This is an API breaking change. Why was it made?

egaxhaj replied:
This is part of the `listenerMx` conversation. ref: https://github.com/cosmos/cosmos-sdk/pull/14207#discussion_r1096560673.

alexanderbez replied:
This breaks the API w/o sufficient reason. I don't understand the convo around concurrency.

egaxhaj replied:
reverted

alexanderbez commented on store/rootmulti/store.go:
```diff
@@ -485,12 +515,14 @@ func (rs *Store) CacheWrapWithTrace(_ io.Writer, _ types.TraceContext) types.Cac
 // CacheMultiStore creates ephemeral branch of the multi-store and returns a CacheMultiStore.
 // It implements the MultiStore interface.
 func (rs *Store) CacheMultiStore() types.CacheMultiStore {
+	rs.listenersMx.Lock()
+	defer rs.listenersMx.Unlock()
```
I don't understand why this has to be synchronized. Where is the concurrency?

egaxhaj replied:
It was synchronized based on review from @peterbourgon (ref: https://github.com/cosmos/cosmos-sdk/pull/14207#discussion_r1084643517). If there is no risk of concurrency then it isn't necessary.

alexanderbez replied:
There is zero concurrency here. I would rather avoid mutexes if possible, especially in a hot code path like this.

egaxhaj replied:
I'll revert the change.

egaxhaj replied:
done

egaxhaj commented on baseapp/abci.go:
```diff
@@ -380,14 +384,18 @@ func (app *BaseApp) CheckTx(req abci.RequestCheckTx) abci.ResponseCheckTx {
-			if err := streamingListener.ListenDeliverTx(app.deliverState.ctx, req, res); err != nil {
-				panic(fmt.Errorf("DeliverTx listening hook failed: %w", err))
+		// call the streaming service hook with the EndBlock messages
+		for _, abciListener := range app.streamingManager.AbciListeners {
```
Fixed.

egaxhaj commented:
@alexanderbez I've reverted the changes and removed the mutex. could use another 👀

tac0turtle commented:
@kocubinski can we get your review here

kocubinski commented on store/listenkv/store.go:
```diff
@@ -35,14 +35,14 @@ func (s *Store) Get(key []byte) []byte {
 	types.AssertValidKey(key)
 	s.parent.Set(key, value)
-	s.onWrite(false, key, value)
+	s.listener.OnWrite(s.parentStoreKey, key, value, false)
```
possible nil reference here right? since `listener` is a pointer type.

egaxhaj replied:
> > > > What is in the 22MB file `store/streaming/abci/examples/file/file`, should this be committed?
> > > 
> > > 
> > > @egaxhaj would be good to get more information on this as well.
> > 
> > 
> > These are the go example plugin binaries (stdout, file). I can remove the `file` binary but the `stdout` binary is required to run the unit tests during CI as it is the default.
> 
> lets remove the file and the binary may be harder.

I've removed the file plugin binary and added a readme and .gitignore in that folder. I've resolved the conflicts as well.

egaxhaj replied:
It is possible only if misconfigured. listenkv stores are attached by checking if `ListeningEnabled` is `true`. `ListeningEnabled` checks to to see if a MemoryListener is `non-nil` for each KeyStore that a listener will be attached. this is a hot code path so let's leave this as is to avoid an extract `if` check for an unlikely case due to misconfiguration.

https://github.com/cosmos/cosmos-sdk/blob/be35b4fdceeaf889471d89f5a4667482ef1aedf1/store/rootmulti/store.go#L416-L421

https://github.com/cosmos/cosmos-sdk/blob/be35b4fdceeaf889471d89f5a4667482ef1aedf1/store/rootmulti/store.go#L520-L522

kocubinski commented:
What is in the 22MB file `store/streaming/abci/examples/file/file`, should this be committed?

kocubinski commented on x/nft/go.mod:
```diff
@@ -149,6 +154,15 @@ require (
 
+// Here are the short-lived replace from the module.
+// Replace here are pending PRs, or version to be tagged.
+replace (
```
Did you try using a go.work file in dev so that we don't end up committing a replace directive like this?

egaxhaj replied:
No. I'm tagging along with what's been pushed to main and updated that.

egaxhaj commented:
@alexanderbez @kocubinski are you good to go with approvals here? Any changes I need to make?

tac0turtle commented:
@egaxhaj could you update to master and then we can merge this pr.

tac0turtle commented:
> What is in the 22MB file `store/streaming/abci/examples/file/file`, should this be committed?

@egaxhaj would be good to get more information on this as well.

kocubinski commented:
Personally I don't think we should be adding so many replace directives to go.mod files if a go.work file would do that job.  Does CI fail without the replace directive?

I also don't understand why we're committing 44 MB of opaque binary files.

tac0turtle commented:
> Does CI fail without the replace directive?

Replace directives is something we do now or we run go get on the version in the pr. This removes the need for the replace tag. Then once tagged dependabot makes the change. We dont commit the go.work file, it is mainly meant for local development. this has caused some confusino

kocubinski commented:
> Replace directives is something we do now or we run go get on the version in the pr. This removes the need for the replace tag. 

What is a "replace tag"?

>We dont commit the go.work file, it is mainly meant for local development. this has caused some confusion

I know what we don't, and also understand that function of a go.work is for local development.  That's why I asked if CI failed if the replace directives are removed.  So, does it?

tac0turtle commented:
> > Replace directives is something we do now or we run go get on the version in the pr. This removes the need for the replace tag.
> 
> What is a "replace tag"?

https://go.dev/ref/mod#go-mod-file-replace

> > We dont commit the go.work file, it is mainly meant for local development. this has caused some confusion
> 
> I know what we don't, and also understand the function of a go.work is for local development. That's why I asked if CI failed if the replace directives are removed. So, does it?

On other prs it can, im not sure about this pr.

egaxhaj commented:
> > > > What is in the 22MB file `store/streaming/abci/examples/file/file`, should this be committed?
> > > 
> > > 
> > > @egaxhaj would be good to get more information on this as well.
> > 
> > 
> > These are the go example plugin binaries (stdout, file). I can remove the `file` binary but the `stdout` binary is required to run the unit tests during CI as it is the default.
> 
> lets remove the file and the binary may be harder.

I've removed the file plugin binary and added a .gitignore and readme. Conflicts have been resolved as well.

egaxhaj commented:
@tac0turtle can we coordinate the merge for this PR? It's hard to keep up with other merges to main. I've had to update the branch 3 times in the last hour.

egaxhaj commented:
> > > Replace directives is something we do now or we run go get on the version in the pr. This removes the need for the replace tag.
> > 
> > 
> > What is a "replace tag"?
> 
> https://go.dev/ref/mod#go-mod-file-replace same thing as replace directive
> 
> > > We dont commit the go.work file, it is mainly meant for local development. this has caused some confusion
> > 
> > 
> > I know what we don't, and also understand the function of a go.work is for local development. That's why I asked if CI failed if the replace directives are removed. So, does it?
> 
> On other prs it can, im not sure about this pr.

I am certain that some tests in CI will fail on this PR. Running `make test` locally fails with compilation errors if the `replace` tags are not in place because the other modules don't currently have the plugin code that resides in the `store` & `baseapp` package.

julienrbrt commented:
🚀 
@egaxhaj do you wanna do the chore (removing the replaces and using a pseudo version from main) as a follow-up or should someone from our team do it?

egaxhaj commented:
Someone from the team please. I'm not familiar with the process the team follows.

julienrbrt commented on baseapp/abci.go:
```diff
@@ -215,10 +215,12 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
+	for _, abciListener := range app.streamingManager.ABCIListeners {
+		ctx := app.deliverState.ctx
+		blockHeight := ctx.BlockHeight()
+		if err := abciListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
```
Curious, doing the chore, and here, why `app.deliverState.ctx` is used and not `ctx`?
